cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/SubstationIdMapping.java

```diff
@@ -10,70 +10,302 @@
-        if (mapping.containsKey(sid)) {
-            return mapping.get(sid);
+        if (substationMapping.containsKey(sid)) {
+            return substationMapping.get(sid);
         }
         return sid;
     }
 
+    // All the keys for a given value, all the merged substations that have cgmesIdentifier as representative
+    public List<String> mergedSubstations(String cgmesIdentifier) {
+        String sid = context.namingStrategy().getId(CgmesNames.SUBSTATION, cgmesIdentifier);
+        return substationMapping.entrySet().stream().filter(record -> record.getValue().equals(sid))
+            .map(Map.Entry::getKey).collect(Collectors.toList());
+    }
+
+    public boolean voltageLevelIsMapped(String cgmesIdentifier) {
+        String vlid = context.namingStrategy().getId(CgmesNames.VOLTAGE_LEVEL, cgmesIdentifier);
+        return voltageLevelMapping.containsKey(vlid);
+    }
+
+    public String voltageLevelIidm(String cgmesIdentifier) {
+        String vlid = context.namingStrategy().getId(CgmesNames.VOLTAGE_LEVEL, cgmesIdentifier);
+        if (voltageLevelMapping.containsKey(vlid)) {
+            return voltageLevelMapping.get(vlid);
+        }
+        return vlid;
+    }
+
+    // All the keys for a given value, all the merged voltageLevels that have cgmesIdentifier as representative
+    public List<String> mergedVoltageLevels(String cgmesIdentifier) {
+        String vlid = context.namingStrategy().getId(CgmesNames.VOLTAGE_LEVEL, cgmesIdentifier);
+        return voltageLevelMapping.entrySet().stream().filter(record -> record.getValue().equals(vlid))
+            .map(Map.Entry::getKey).collect(Collectors.toList());
+    }
+
     public void build() {
-        // CGMES standard:
-        // "a PowerTransformer is contained in one Substation but it can connect a Terminal to
-        // another Substation"
-        // Ends of transformers need to be in the same substation in the IIDM model.
-        // We will map some CGMES substations to a single IIDM substation
-        // when they are connected by transformers,
-        // that is, when there are at least one power transformer that has terminals on both
-        // substations
-
-        Graph<String, Object> g = graphSubstationsTransformers();
-        new ConnectivityInspector<>(g).connectedSets().stream()
-                .filter(substationIds -> substationIds.size() > 1)
-                .forEach(substationIds -> {
-                    String selectedSubstationId = representativeSubstationId(substationIds);
-                    for (String substationId : substationIds) {
-                        if (!substationId.equals(selectedSubstationId)) {
-                            mapping.put(substationId, selectedSubstationId);
-                        }
+        Map<String, List<String>> voltageLevelAdjacency = new HashMap<>();
+        Map<String, List<String>> substationAdjacency = new HashMap<>();
```

Comment:
From my point of view, to avoid duplicates we should use `Map<String, Set<String>>`

Reply:
Done